hrc_speech_prediction/tree.py

```diff
@@ -0,0 +1,248 @@
+
+    def take_action(self, model="both", utter=None,
+                    plot=False, return_probs=False):
+
+        context_probs = self.get_context_probs()
+        if model == "context":
+            probs = self.get_context_probs()
+        elif model == "speech" and utter is not None:
+            probs = self.get_speech_probs(utter)
+        elif model == "both" and utter is not None:
+            context_probs = self.get_context_probs()
+            speech_probs = self.get_speech_probs(utter)
+
+            probs = np.multiply(context_probs, speech_probs)
+
+        action = self.pred_action(probs)
+        self.curr = self.curr._get_next_node(action)
+
+        if plot:
+            self.plot_predicitions(speech_probs, context_probs,
+                                   combined_probs, utter)
+        if return_probs and model=="both":
+            return action, speech_probs, context_probs, probs
+        else:
+            return action, probs
+
+    def reset(self):
+        self._curr = self.root
+
+    def plot_predicitions(self, speech, context, both, utter,
+                          actual=None, save_path=None):
+        "Plots the probabilities for each possible action provided by speech, \
+        context, and speech + context "
+        X = np.arange(len(both))
+        fig, ax = plt.subplots(nrows=1, ncols=1)
+
+        # Want to normalize 'both' probs for easier visual comparison
+        nrmlz = 1.0 / sum(both)
+
+        ax.bar(X-0.2, speech, width=0.2, color='r', align='center')
+        ax.bar(X, context, width=0.2, color='b', align='center')
+        ax.bar(X+0.2, both * nrmlz, width=0.2, color='g', align='center')
+
+        ax.legend(('Speech', 'Context', 'Both'))
+
+        rects = ax.patches
+        max_prob = max(both * nrmlz)
+
+        #This draws a star above most probable action
+        for r in rects:
+            if r.get_height() == max_prob:
+                ax.text(r.get_x() + r.get_width()/2,
+                        r.get_height() * 1.01, '*', ha='center', va='bottom')
+
+        if actual:
+            ax.text(self.speech_model.actions.index(actual), max_prob, "$")
+
+        plt.xticks(X, self.speech_model.actions, rotation=70)
+        plt.title(utter)
+
+        if save_path:
+            plt.savefig(save_path)
+        else:
+            plt.show()
+
+    def __str__(self):
+        return self.root.__str__()
+
+
+class Node(object):
+    def __init__(self):
+        "A state in a task trajectory"
+        self._count = 1.0
+        self._children = {}  # keys: action taken, val: list of nodes
+
+
+    @property
+    def n_children(self):
+        return len(self._children)
+
+    @property
+    def seen_children(self):
+        return self._children.keys()
+
+    @property
+    def children_counts(self):
+        return [self._children[c]._count for c in self.seen_children]
+
+    def _increment_count(self):
+        self._count += 1.0
+        return self
+
+    def add_branch(self, new_acts):
```

Comment:
It seems that for the `root` the count is never incremented. Since it is never used it does not really matter, but on the other side, instead of incrementing the child from the parent you could increment the child when calling `add_branch` on it. If you do so, you would have to initialize the node count to `0.` when creating the object.